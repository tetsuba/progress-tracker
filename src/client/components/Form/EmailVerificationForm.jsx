import React, { useState } from 'react'
import { Link } from 'react-router-dom'
import { Button, Form, Col } from 'react-bootstrap'

// HOOKS
import { useInputChange } from '../../hooks/hooks'

// COMPONENTS
import Box from '../Box/Box'

// ICONS
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCaretLeft } from '@fortawesome/free-solid-svg-icons'

export default function EmailVerificationForm(props) {
  const { handleSubmit, error } = props
  const [errorMessage] = useState(() => error && error.graphQLErrors[0].message)
  const [inputs, setInputs] = useInputChange({ email: '' })
  const options = { variables: { input: inputs } }

  return (
    <Box max={500}>
      <Col>
        <h3 className="text-danger">Could not sign you in</h3>
        <p>
          Your email address has not been confirmed. Please enter your email
          address below and a link will be sent to confirm your email address.
        </p>
        <Form
          id="EmailVerificationForm"
          onSubmit={(e) => {
            e.preventDefault()
            handleSubmit(options)
          }}
        >
          <Form.Group controlId="EmailVerificationEmail">
            <Form.Control
              required
              type="email"
              placeholder="email@address.com"
              name="email"
              onChange={setInputs}
              value={inputs.email}
              isInvalid={!!errorMessage}
            />
            <Form.Control.Feedback type="invalid">
              {errorMessage}
            </Form.Control.Feedback>
            <Form.Text className="text-muted">
              We'll never share your email with anyone else.
            </Form.Text>
          </Form.Group>
          <Button
            id="EmailVerificationSubmit"
            className="float-right"
            variant="primary"
            type="submit">
            Send
          </Button>
        </Form>
        <Link to="/">
          <FontAwesomeIcon icon={faCaretLeft} /> back
        </Link>
      </Col>
    </Box>
  )
}
