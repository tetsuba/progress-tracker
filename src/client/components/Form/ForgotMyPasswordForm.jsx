import React from 'react';
import { useMutation } from '@apollo/react-hooks';
import { Button, Form, Row, Col } from 'react-bootstrap';

// MUTATIONS
import { SEND_PASSWORD_RESET_CONFIRMATION_MUTATION } from '../../api/user/user.mutation';

// ICONS
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faCaretLeft} from "@fortawesome/free-solid-svg-icons";

// COMPONENT
import TextLink from "../TextLink/TextLink";
import {useError, useInputChange, useSuccess} from "../../hooks/hooks";

const ForgotMyPasswordForm = ({ defaultEmail = '', resetPassword}) => {
    const [ sendPasswordResetConfirmation ] = useMutation(SEND_PASSWORD_RESET_CONFIRMATION_MUTATION);
    const [inputs, setInputs] = useInputChange({ email: defaultEmail });
    const [error, setError] = useError({ message: '' });
    const [success, setSuccess] = useSuccess('sendPasswordResetConfirmation');
    const options = { variables: { input: inputs } };

    return success
        ? (
            <Row>
                <h3>Please check your email.</h3>
            </Row>
        )
        :(
            <Col>
                <Row>
                    <h3>Forgot my password</h3>
                </Row>
                <Row>
                    <p>Enter your email address and we will send you a link to reset your password.</p>
                </Row>
                <Row>
                    <Form
                        id="ForgotMyPasswordForm"
                        className="w-100"
                        onSubmit={(e) => {
                            e.preventDefault();
                            sendPasswordResetConfirmation(options)
                                .then(setSuccess)
                                .catch(setError)
                        }}
                    >
                        <Form.Group controlId="forgotMyPasswordEmail">
                            <Form.Control
                                required
                                type="email"
                                placeholder="Enter email"
                                name="email"
                                onChange={setInputs}
                                value={inputs.email}
                                isInvalid={!!error.message}
                            />
                            <Form.Control.Feedback type="invalid">
                                {error.message}
                            </Form.Control.Feedback>
                            <Form.Text className="text-muted">
                                We'll never share your email with anyone else.
                            </Form.Text>
                        </Form.Group>
                        <Button id="ForgotMyPasswordSubmit" variant="primary" type="submit">
                            Reset Password
                        </Button>
                    </Form>
                </Row>
                { resetPassword && (
                    <Row className="mt-3">
                        <TextLink eventHandler={resetPassword}>
                            <FontAwesomeIcon icon={faCaretLeft} /> back
                        </TextLink>
                    </Row>
                )}

            </Col>
        )
};

export default ForgotMyPasswordForm;
