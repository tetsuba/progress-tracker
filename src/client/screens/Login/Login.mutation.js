import gql from 'graphql-tag';

export const LOGIN_MUTATION = gql`
    mutation($input: LoginUserInput!) {
        userLogin(input: $input) {
            id
            firstName
            lastName
            token
            email
        }
    }
`;
