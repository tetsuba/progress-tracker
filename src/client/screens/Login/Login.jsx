import React, { useState, useContext } from 'react';
import { useMutation } from '@apollo/react-hooks';
import { Container, Row } from 'react-bootstrap';
import { UserContext } from '../../context/UserContext';

// CONTEXT
import { AuthenticatedContext } from '../../context/AuthenticatedContext';

// MUTATION
import { LOGIN_MUTATION } from '../../api/user/user.mutation';

// COMPONENTS
import LoginForm from '../../components/Form/LoginForm';
import EmailPasswordResetForm from '../../components/Form/EmailPasswordResetForm';

// UTILS
import { getLoginStatus } from "../../components/Form/form-utils";
import { useLocalStorage } from "../../hooks/hooks";
import { TOKEN_KEY } from "../../const/localStorage";

const Login = () => {
    const [ userLogin, { loading: mutationLoading }] = useMutation(LOGIN_MUTATION);
    const { toggle: authenticateUser } = useContext(AuthenticatedContext);
    const { userLoggedIn } = useContext(UserContext);
    const [error, setError] = useState({});
    const [resetPassword, setResetPassword] = useState(false);
    const [setLocalStorage] = useLocalStorage(TOKEN_KEY);

    const handleSubmit = (e, inputs) => {
        e.preventDefault();

        userLogin({ variables: { input: inputs } })
            .then(({ data }) => {
                console.log(data);
                userLoggedIn(data.userLogin);
                authenticateUser(true);
                setLocalStorage(data.userLogin.token);
            })
            .catch((err) => {
                // TODO: rename to not be the same as the state error name
                const error = err.graphQLErrors[0];

                if(error.name === 'email_not_verified') {
                    setError({
                        ...error,
                        emailNotVerified: true,
                    });
                } else {
                    setError({
                        email: error,
                        password: error,
                    })
                }
            })
    };

    return (
        <Container>
            <Row className="mt-5">
                {{
                    loading: (<div>Loading</div>),
                    form: (
                        <LoginForm
                            error={error}
                            handleSubmit={handleSubmit}
                            resetPassword={() => setResetPassword(true)}
                        />
                    ),
                    reset: (
                        <EmailPasswordResetForm
                            resetPassword={() => setResetPassword(false)}
                        />
                    ),
                }[getLoginStatus(mutationLoading, resetPassword)]}
            </Row>
        </Container>
    )
};

export default Login;
