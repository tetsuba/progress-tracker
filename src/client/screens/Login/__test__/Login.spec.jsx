import React from 'react'
import { act } from 'react-dom/test-utils'
import {
  delay,
  graphRenderer,
  updateTextInput,
} from '../../../../test/testHelper'

jest.mock('react-router-dom', () => ({
  Link: ({ children }) => children,
}))

// COMPONENTS
import Login from '../Login'
import ForgotMyPasswordForm from '../../../components/Form/ForgotMyPasswordForm'
import EmailVerificationForm from '../../../components/Form/EmailVerificationForm'

// MOCK DATA
import {
  loginMockDataErrorEmailNotVerified,
  loginMockDataSuccess,
  verifyUserEmailMutationSuccess,
  passwordResetMockDataSuccess,
} from '../../../../test/mockApi/user/userMockMutation'

describe('<Login>', () => {
  describe('Initial render', () => {
    it('should render "login"', async () => {
      let wrapper
      act(() => {
        wrapper = graphRenderer(Login, [], {})
      })
      expect(wrapper.find(Login)).toMatchSnapshot()
    })
    describe('A user logs-in with a valid email and password', () => {
      it('should save a token in localStorage', async () => {
        let wrapper
        await act(async () => {
          wrapper = graphRenderer(Login, [loginMockDataSuccess], {})
          await delay()
        })

        expect(window.localStorage.getItem('ptToken')).toBeNull()
        wrapper.update()

        act(() => {
          updateTextInput(wrapper, 'email', 'test@test.com')
        })

        wrapper.update()

        act(() => {
          updateTextInput(wrapper, 'password', '1234')
        })

        wrapper.update()

        await act(async () => {
          await wrapper.find('#LoginForm').get(0).props.onSubmit({
            preventDefault: jest.fn(),
          })
        })

        wrapper.update()
        expect(window.localStorage.getItem('ptToken')).toEqual('token')
      })
    })

    describe('A user logs-in with an email address that is not verified', () => {
      /* NOTE:
       * After completing the registration form a user is sent a
       * confirmation email to verify the users email address.
       * If the user tries to login without confirming their email
       * address. The user will be directed to this component to send
       * another confirmation.
       * */

      let wrapper
      beforeEach(async () => {
        await act(async () => {
          wrapper = graphRenderer(
            Login,
            [
              loginMockDataErrorEmailNotVerified,
              verifyUserEmailMutationSuccess,
            ],
            {}
          )
          await delay()
        })

        wrapper.update()

        act(() => {
          updateTextInput(wrapper, 'email', 'test@test.com')
        })

        wrapper.update()

        act(() => {
          updateTextInput(wrapper, 'password', '1234')
        })

        wrapper.update()

        await act(async () => {
          await wrapper.find('#LoginForm').get(0).props.onSubmit({
            preventDefault: jest.fn(),
          })
        })

        wrapper.update()
      })

      it('should render "emailNotVerified" form', async () => {
        expect(wrapper.find(EmailVerificationForm)).toMatchSnapshot()
      })
      describe('A user submits a valid email address', () => {
        it('should render "success" message', async () => {
          act(() => {
            updateTextInput(wrapper, 'email', 'test@test.com')
          })

          wrapper.update()

          await act(async () => {
            wrapper.find('#EmailVerificationForm').get(0).props.onSubmit({
              preventDefault: jest.fn(),
            })
            await delay()
          })
          wrapper.update()
          expect(wrapper.find(Login)).toMatchSnapshot()
        })
      })
    })

    describe('A user clicks on "forgot password" link', () => {
      let wrapper
      beforeEach(async () => {
        act(() => {
          wrapper = graphRenderer(
            Login,
            [loginMockDataSuccess, passwordResetMockDataSuccess],
            {}
          )
        })
        wrapper.find('#TextLink').simulate('click')
        wrapper.update()
      })

      it('should render "forgotMyPassword" form', async () => {
        expect(wrapper.find(ForgotMyPasswordForm)).toMatchSnapshot()
      })
      describe('A user submits a valid email address', () => {
        it('should render "success"', async () => {
          act(() => {
            updateTextInput(wrapper, 'email', 'test@test.com')
          })

          wrapper.update()

          await act(async () => {
            await wrapper.find('#ForgotMyPasswordForm').get(0).props.onSubmit({
              preventDefault: jest.fn(),
            })
            await delay()
          })

          wrapper.update()
          expect(wrapper.find(Login)).toMatchSnapshot()
        })
      })
    })
  })
})
