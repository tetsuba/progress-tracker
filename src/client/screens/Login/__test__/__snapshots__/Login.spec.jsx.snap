// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login> @Render should render "emailNotVerified" 1`] = `
<Container
  className="pt-5"
  fluid={false}
>
  <Row
    noGutters={false}
  >
    <EmailVerificationForm
      setPageState={[Function]}
    >
      <h3
        className="text-danger"
      >
        Could not sign you in
      </h3>
      <p>
        Your email address has not been confirmed. Please enter your email address below and a link will be sent to confirm your email address.
      </p>
    </EmailVerificationForm>
  </Row>
</Container>
`;

exports[`<Login> @Render should render "forgetMyPassword" 1`] = `
<Container
  className="pt-5"
  fluid={false}
>
  <Row
    noGutters={false}
  >
    <ForgotMyPasswordForm
      setPageState={[Function]}
    >
      <h3>
        Forgot my password
      </h3>
      <p>
        Enter your email address and we will send you a link to reset your password.
      </p>
    </ForgotMyPasswordForm>
  </Row>
</Container>
`;

exports[`<Login> @Render should render "login" 1`] = `
<Container
  className="pt-5"
  fluid={false}
>
  <Row
    noGutters={false}
  >
    <LoginForm
      setPageState={[Function]}
    />
  </Row>
</Container>
`;

exports[`<Login> @Render should render "success" 1`] = `
<Container
  className="pt-5"
  fluid={false}
>
  <Row
    noGutters={false}
  >
    <Col
      id="LoginSuccess"
    >
      <h3>
        Please check your email.
      </h3>
    </Col>
  </Row>
</Container>
`;
